dask/dataframe/core.py

```diff
@@ -4132,6 +4171,33 @@ def applymap(self, func, meta="__no_default__"):
     def round(self, decimals=0):
         return elemwise(M.round, self, decimals)
 
+    @derived_from(pd.DataFrame)
+    def mode(self, axis=0, numeric_only=False, dropna=True, split_every=False):
+        axis = self._validate_axis(axis)
+        if axis == 1:
+            return super(DataFrame, self).mode(
+                axis=axis, dropna=dropna, split_every=split_every
+            )
+        elif axis == 0:
+            from .multi import concat
+
+            mode_series_list = []
+            for col in self.columns:
+                mode_series_list.append(
+                    super(Series, self[col]).mode(
+                        dropna=dropna, split_every=split_every
+                    )
+                )
+
+            ddf = concat(mode_series_list, axis=1)
```

Comment:
@TomAugspurger Thank you for your response.  I wasn't aware of value_counts, but it promises to simplify things quite a bit.  I'm currently implementing mode using value_counts, but I'm hitting an error here on `ddf = concat(mode_series_list, axis=1)`.  The error is below.
```python3  
File "/home/balast/PycharmProjects/dask/dask_repo/dask/dataframe/multi.py", line 865, in concat_and_check
    raise ValueError("Concatenated DataFrames of different lengths")
ValueError: Concatenated DataFrames of different lengths
```
There is no error when concatenating series of different lengths if the divisions are known, but the divisions aren't known here. I believe I can get around this using the delayed api similar to what I did in the last PR, where I add np.nans to each short array until they are all the same length, but before I do so, I was wondering if you have a suggestion to concatenate series of various lengths without using the delayed api.  Things will be simpler if it's possible.

Reply:
I don't think you should be concatenating on axis=1. The idea is to define mode in terms of a value_counts over the entire dataset followed by a selection of the index values where the count equals the max. Does that make sense?

So I don't think the lines https://github.com/dask/dask/pull/5958/files#diff-492da7893fa5fa6ad3a6ef6cdef985f2R4184-R4190 are wanted.

Reply:
> The idea is to define mode in terms of a value_counts over the entire dataset followed by a selection of the index values where the count equals the max.

To be consistent with pandas, the mode is calculated along each row or column of the dataframe (depending on the value of axis).  So in the current implementation _Frame.mode does what you describe for a single series.  Then DataFrame.mode() merely calls _Frame.mode for each series in itself (when axis=1).  

The mode of each series can be a different number of values so each series is of different length.  I'm getting an error when trying to concatenate the various length series together into a single dataframe.

Example of Pandas Functionality
-------------------------------------------
```python3 
import pandas as pd 
df = pd.DataFrame({'a': [1,1,2,3], 'b': [2,2,3,3], 'c': [3,4,5,6]})
mode_result = df.mode(axis=0) 
```
![image](https://user-images.githubusercontent.com/23342526/76124601-b0a22f80-5fc0-11ea-8593-9fd58f903ae8.png)

Reply:
Gotcha. I'm not sure what's best then.

Reply:
No worries, I think I can manage with the delayed api

Reply:
I made another PR https://github.com/dask/dask/pull/5990 that will make this possible without resorting to the delayed API.  All tests pass on that PR.

Reply:
The other PR proved to be more complex than anticipated.  I closed it, and instead added to this PR to produce the concatenation appropriately for mode.